.azure/scripts/setup-kind.sh

```diff
@@ -120,19 +196,116 @@ function adjust_inotify_limits {
+            fi
+            $DOCKER_CMD network connect "${network_name}" "${registry_name}"
+        else
+            $DOCKER_CMD network connect "${network_name}" "${registry_name}"
+        fi
+    fi
+}
+
+: '
+@brief: Creates a KIND cluster configuration and provisions the cluster.
+@param:
+        1) control_planes - Number of control-plane nodes.
+@global:
+        KIND_CLUSTER_NAME - Name of the KIND cluster.
+        KIND_NODE_IMAGE   - KIND node image to use.
+        IP_FAMILY         - Determines whether to configure IPv4 or IPv6 settings
+@note: Writes the cluster configuration to a temporary file before creating the cluster.
+'
+function create_kind_cluster {
+    local control_planes="$1"
+
+# Start the cluster configuration
+    cat <<EOF > /tmp/kind-config.yaml
+kind: Cluster
+apiVersion: kind.x-k8s.io/v1alpha4
+nodes:
+EOF
+
+    # Add control-plane nodes
+    for i in $(seq 1 "$control_planes"); do
+        echo "    - role: control-plane" >> /tmp/kind-config.yaml
+    done
+
+    # Add worker nodes
+    cat <<EOF >> /tmp/kind-config.yaml
+    - role: worker
+    - role: worker
+    - role: worker
+EOF
+
+    # Add specific containerd configuration for IPv4/IPv6
+    if [[ "$IP_FAMILY" == "ipv6" ]]; then
+        cat <<EOF >> /tmp/kind-config.yaml
+containerdConfigPatches:
+- |-
+  [plugins."io.containerd.grpc.v1.cri".registry.mirrors."myregistry.local:5001"]
+      endpoint = ["http://myregistry.local:5001"]
+EOF
+    else
+        cat <<EOF >> /tmp/kind-config.yaml
+containerdConfigPatches:
+- |-
+  [plugins."io.containerd.grpc.v1.cri".registry]
+      config_path = "/etc/containerd/certs.d"
+EOF
+    fi
+
+    # Add networking configuration
+    cat <<EOF >> /tmp/kind-config.yaml
+networking:
+  ipFamily: $IP_FAMILY
+EOF
+
+
+    # Create the KIND cluster
+    kind create cluster \
+        --image "$KIND_NODE_IMAGE" \
+        --name "$KIND_CLUSTER_NAME" \
+        --config=/tmp/kind-config.yaml
+
+    echo "KIND cluster '${KIND_CLUSTER_NAME}' created successfully with IP family '${IP_FAMILY}'."
+}
+
 setup_kube_directory
 install_kubectl
 install_kubernetes_provisioner
 adjust_inotify_limits
+load_ip6_tables_module_to_kernel
 
 reg_name='kind-registry'
 reg_port='5001'
+# by default using podman we have to use single control-plane because of https://github.com/kubernetes-sigs/kind/issues/2858
+control_planes=1
+
+if is_docker; then
+    control_planes=3
+fi
 
 if [[ "$IP_FAMILY" = "ipv4" || "$IP_FAMILY" = "dual" ]]; then
     hostname=$(hostname --ip-address | grep -oE '\b([0-9]{1,3}\.){3}[0-9]{1,3}\b' | awk '$1 != "127.0.0.1" { print $1 }' | head -1)
 
     # update insecure registries
-    updateDockerDaemonConfiguration "{ \"insecure-registries\" : [\"${hostname}:${reg_port}\"] }"
+    updateContainerRuntimeConfiguration "${hostname}:${reg_port}"
 
     # Create kind cluster with containerd registry config dir enabled
     # TODO: kind will eventually enable this by default and this patch will
```

Comment:
is this needed now?

Reply:
I still see https://github.com/kubernetes-sigs/kind/issues/2875 to be open. So I think yes.